### PR TITLE
Add page that shows problem history for a single user and course

### PR DESCRIPTION
**Pull request by itsjeyd**
This PR adds a new page to the Insights dashboard that shows, for a specific user and course, the list of problems that the user has attempted, along with the total number of attempts and the final (i.e., most recent) score.

The new page can be accessed by choosing the "Problem Data" entry from the dropdown menu that is available in the navigation bar while viewing a user's profile. (The dropdown menu also provides a "User Profile" entry that allows to switch back to a user's profile after viewing problem information on the new page.)

While viewing problem data the dashboard is in "user mode": The course mode selector can be used to enable course mode and jump back to the home page of the parent course; the user mode selector points to the list of all users currently enrolled in the parent course. (See https://github.com/edx/edx-analytics-dashboard/pull/338 for more information about mode selectors, user mode, and course mode.)
### Screenshots

New page listing problems user has attempted:

user-problem-history

The dashboard uses the Course Structure API to obtain display names for problems when generating this page. If the `course_api_enabled` switch is off, it falls back on problem IDs:

user-problem-history-course-api-disabled

Navigation lens on new page:

navigation-lens-on-problem-history-page

Navigation lens on profile page:

navigation-lens-on-profile-page
### Dependencies
- https://github.com/edx/edx-analytics-pipeline/pull/147
- https://github.com/edx/edx-analytics-data-api/pull/81
- https://github.com/edx/edx-analytics-data-api-client/pull/22
- https://github.com/edx/edx-analytics-dashboard/pull/338

Note that https://github.com/edx/edx-analytics-dashboard/pull/338 has data-api and data-api-client dependencies of its own. For the purpose of testing this PR you will therefore need to work with branches that contain relevant changes for both https://github.com/edx/edx-analytics-dashboard/pull/338 and this PR. You won't need to create these branches yourself; they have already been set up for you (cf. test instructions below).
### Testing
1. Test https://github.com/edx/edx-analytics-pipeline/pull/147 and confirm it's working.
2. Prepare the **data-api** repository by switching to [this branch](https://github.com/open-craft/edx-analytics-data-api/tree/user-problem-data) (`user-problem-data`).
3. Prepare the **data-api-client** repository by switching to [this branch](https://github.com/open-craft/edx-analytics-data-api-client/tree/user-problem-data) (`user-problem-data`). 
4. Install the updated data API client into the dashboard's virtualenv. Use these commands (as `analytics` user):
   
   ```
   cd ~/apps/data-api-client/
   deactivate
   source ~/venvs/dashboard/bin/activate
   pip install -e .
   ```
5. Start the data API server (as `analytics` user):
   
   ```
   cd ~/apps/data-api/ && ./manage.py runserver 0.0.0.0:9001
   ```
6. Start the dashboard (as `analytics` user):
   
   ```
   cd ~/apps/dashboard/ && ./manage.py runserver 0.0.0.0:9999
   ```
7. Access the Insights page for the demo course at http://192.168.33.11:9999/courses/edX/DemoX/Demo_Course/.
8. Follow these steps to try out the new functionality:
   - Click the user mode selector to switch to user mode and access the list of enrolled students.
   - Click the username of a student to access their profile.
   - Expand the dropdown menu in the navigation bar and click "Problem Data" to access the new page.
### Partner Information

Not an edX partner - 3rd party-hosted open edX instance

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @itsjeyd! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=343&repo=edx%2Fedx-analytics-dashboard

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @itsjeyd! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-analytics-dashboard&number=343

**Comment by bradenmacdonald** [comment]
Please fix the failing Travis build. Otherwise this is :+1: from me :)

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @itsjeyd! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=343&repo=edx%2Fedx-analytics-dashboard

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @itsjeyd! I've created [OSPR-790](https://openedx.atlassian.net/browse/OSPR-790) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).
We can't start reviewing your pull request until you've   added yourself to the [AUTHORS](https://github.com/edx/edx-analytics-dashboard/blob/master/AUTHORS) file . Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

File: analytics_dashboard/courses/templates/courses/home.html [hunk 1]
```diff
@@ -17,13 +17,32 @@
...
-              {{ item.label }}
-            </span></a>
+                {{ item.label }}
+              </span>
+            </a>
           </li>
         {% endfor %}
       </ul>
+      <ul id="mode-selectors" class="nav navbar-nav navbar-right">
+        <li class="active nav-section has-tooltip" data-toggle="tooltip" title="You're in course mode">
```
Comment (sarina) [inline analytics_dashboard/courses/templates/courses/home.html]: i18n

also this string will be super unclear out of context

File: analytics_dashboard/courses/templates/courses/home.html [hunk 2]
```diff
@@ -17,13 +17,32 @@
...
+          <a href="#">
+            <span class="link-label">
+              <i class="ico fa fa-users" aria-hidden="true"></i>
+            </span>
+          </a>
+          <span class="sr-only">({% trans "Active" %})</span>
+        </li>
+        <li class="has-tooltip" data-toggle="tooltip" title="Click to switch to user mode">
```
Comment (sarina) [inline analytics_dashboard/courses/templates/courses/home.html]: i18n

File: analytics_dashboard/courses/views/__init__.py [hunk 3]
```diff
@@ -56,13 +66,41 @@ def dispatch(self, request, *args, **kwargs):
...
+        """
+        key = self._course_structure_cache_key(course_id)
+        structure = cache.get(key)
+
+        if not structure:
+            try:
+                logger.debug("Retrieving structure for course: %s", course_id)
+                structure = self.course_api_client.course_structures(course_id).get()
+                cache.set(key, structure)
+            except HttpClientError as e:
```
Comment (sarina) [inline analytics_dashboard/courses/views/__init__.py]: variables should be at least 3 letters (so `as err` would work)

File: analytics_dashboard/templates/lens-navigation.html [hunk 4]
```diff
@@ -33,6 +33,24 @@
 
         {% include "submenu_navigation.html" with nav_items=secondary_nav_items%}
       </ul>
+
+      <ul id="mode-selectors" class="nav navbar-nav navbar-right">
+        <li class="active nav-section has-tooltip" data-toggle="tooltip" title="You're in course mode">
```
Comment (sarina) [inline analytics_dashboard/templates/lens-navigation.html]: i18n + comment

File: analytics_dashboard/templates/lens-navigation.html [hunk 5]
```diff
@@ -33,6 +33,24 @@
...
+          <a href="#">
+            <span class="link-label">
+              <i class="ico fa fa-users" aria-hidden="true"></i>
+            </span>
+          </a>
+          <span class="sr-only">({% trans "Active" %})</span>
+        </li>
+        <li class="has-tooltip" data-toggle="tooltip" title="Click to switch to user mode">
```
Comment (sarina) [inline analytics_dashboard/templates/lens-navigation.html]: i18n

File: analytics_dashboard/users/views/base.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+    def dispatch(self, request, *args, **kwargs):
+        self.client = Client(
+            base_url=settings.DATA_API_URL,
+            auth_token=settings.DATA_API_AUTH_TOKEN,
+            timeout=5
+        )
+        try:
+            return super(UsersView, self).dispatch(request, *args, **kwargs)
+        except ClientError as e:
```
Comment (sarina) [inline analytics_dashboard/users/views/base.py]: as err

File: analytics_dashboard/users/views/problem_data.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
+from .base import UsersView
+from .user import SingleUserNavbarMixin
+
+
+class UserProblemDataView(UsersView, SingleUserNavbarMixin):
```
Comment (sarina) [inline analytics_dashboard/users/views/problem_data.py:9]: Docstring? Code comments? File level docstring? Explain yo'self!

**Comment by itsjeyd** [comment]
@sarina Thanks for the comments, I added some commits that address them.

**Comment by sarina** [comment]
@itsjeyd great, thanks! My small concerns are addressed, but the Analytics team overall will need to review this.

**Comment by openedx-webhooks** [comment]
Hello @itsjeyd: We are unable to continue with review of your submission at this time. Please see the associated JIRA ticket for more explanation.

**Pull request closed without merging**